File: python/aifinpay/facilitators/detect.py
"""Pick the right facilitator adapter for a given 402 response."""
from __future__ import annotations

from typing import Optional

import requests

from ..errors import UnsupportedFacilitatorError
from .aifinpay import AiFinPayFacilitator
from .base import Facilitator
from .coinbase import CoinbaseX402Facilitator


# Order matters: most-specific detector first. A response that matches
# AiFinPay's body schema is *also* technically a 402, so we try AiFinPay
# before falling back to the generic header-based check.
REGISTERED: list[type[Facilitator]] = [
    AiFinPayFacilitator,
    CoinbaseX402Facilitator,
]


def detect_facilitator(
    resp: requests.Response,
    override: str = "auto",
) -> Facilitator:
    """Return the facilitator adapter for `resp`.

    Args:
        resp: the 402 response to inspect.
        override: `"auto"` runs detection. Any other string forces that
            facilitator by name (e.g. `"aifinpay"`).

    Raises:
        UnsupportedFacilitatorError: no registered facilitator matched.
    """
    if override and override != "auto":
        for cls in REGISTERED:
            if cls.name == override:
                return cls()  # type: ignore[abstract]
        raise UnsupportedFacilitatorError(
            f"unknown facilitator override: {override!r}. "
            f"known: {[c.name for c in REGISTERED]}"
        )

    for cls in REGISTERED:
        if cls.detect(resp):
            return cls()  # type: ignore[abstract]

    raise UnsupportedFacilitatorError(
        f"402 response did not match any known facilitator. "
        f"Status: {resp.status_code}. "
        f"Headers: {list(resp.headers.keys())[:8]}. "
        f"Body preview: {resp.text[:200]!r}"
    )

File: node/src/facilitators/detect.ts
import { UnsupportedFacilitatorError } from "../errors.js";
import { AiFinPayFacilitator } from "./aifinpay.js";
import type { Facilitator, FacilitatorClass } from "./base.js";
import { CoinbaseX402Facilitator } from "./coinbase.js";
import { StandardX402Facilitator } from "./standard-x402.js";

/**
 * Order matters: most-specific detector first. A response that matches
 * AiFinPay's body schema is also technically a 402, so we try AiFinPay
 * before falling back to header-based checks.
 */
export const REGISTERED: FacilitatorClass[] = [
  AiFinPayFacilitator,
  StandardX402Facilitator,
  CoinbaseX402Facilitator,
];

const BY_NAME = new Map<string, FacilitatorClass>(
  REGISTERED.map((cls) => [cls.name, cls]),
);

export async function detectFacilitator(
  resp: Response,
  override: string = "auto",
): Promise<Facilitator> {
  if (override && override !== "auto") {
    const cls = BY_NAME.get(override);
    if (!cls) {
      throw new UnsupportedFacilitatorError(
        `unknown facilitator override: '${override}'. ` +
          `known: ${[...BY_NAME.keys()].join(", ")}`,
      );
    }
    return new cls();
  }

  for (const cls of REGISTERED) {
    if (await cls.detect(resp)) return new cls();
  }

  const headerKeys: string[] = [];
  resp.headers.forEach((_, key) => headerKeys.push(key));
  throw new UnsupportedFacilitatorError(
    `402 response did not match any known facilitator. ` +
      `Status: ${resp.status}. ` +
      `Headers: ${headerKeys.slice(0, 8).join(", ")}.`,
  );
}
